File: api/Kdkdjdk.js
import express from "express";
import axios from "axios";
import cheerio from "cheerio";
import { Octokit } from "@octokit/rest";
import dotenv from "dotenv";

dotenv.config();

const app = express();

const GITHUB_REPO = process.env.GITHUB_REPO; // e.g. "your-username/your-repo"
const GITHUB_FILE_PATH = "reference_versions.json";
const DISCORD_WEBHOOK_URL = process.env.DISCORD_WEBHOOK_URL;

if (!GITHUB_REPO) {
  console.error("Error: GITHUB_REPO env var is not set");
  process.exit(1);
}

const [owner, repo] = GITHUB_REPO.split("/");

const octokit = new Octokit({
  auth: process.env.GITHUB_TOKEN,
});

const TARGET_URL = "https://executors.samrat.lol/";

async function fetchCurrentVersions() {
  const { data } = await axios.get(TARGET_URL);
  const $ = cheerio.load(data);

  const executors = {};
  $(".executor-card").each((_, el) => {
    const name = $(el).find("h3").text().trim();
    const versionText = $(el).find("span.version").text().trim();
    const version = versionText.replace("Version: ", "");
    if (name && version) {
      executors[name] = version;
    }
  });

  return executors;
}

async function getReferenceVersions() {
  try {
    const response = await octokit.repos.getContent({
      owner,
      repo,
      path: GITHUB_FILE_PATH,
      ref: "main",
    });

    const content = Buffer.from(response.data.content, "base64").toString();
    return JSON.parse(content);
  } catch (err) {
    if (err.status === 404) {
      return {};
    }
    throw err;
  }
}

async function saveReferenceVersions(data) {
  let sha;
  try {
    const response = await octokit.repos.getContent({
      owner,
      repo,
      path: GITHUB_FILE_PATH,
      ref: "main",
    });
    sha = response.data.sha;
  } catch (err) {
    if (err.status !== 404) throw err;
  }

  const content = Buffer.from(JSON.stringify(data, null, 4)).toString("base64");

  await octokit.repos.createOrUpdateFileContents({
    owner,
    repo,
    path: GITHUB_FILE_PATH,
    message: `Update reference_versions.json at ${new Date().toISOString()}`,
    content,
    sha,
    branch: "main",
  });
}

async function notifyDiscord(changes) {
  if (!DISCORD_WEBHOOK_URL) {
    console.log("No Discord webhook URL set. Skipping notification.");
    return;
  }

  const content = changes.join("\n");
  try {
    await axios.post(DISCORD_WEBHOOK_URL, {
      username: "Version Monitor",
      content: `**Executor Version Changes Detected:**\n${content}`,
    });
  } catch (err) {
    console.error(
      "Failed to send Discord notification:",
      err.response?.status,
      err.response?.data || err.message
    );
  }
}

app.get("/check", async (req, res) => {
  try {
    const reference = await getReferenceVersions();
    const current = await fetchCurrentVersions();

    const changesDetected = [];

    for (const [name, currVersion] of Object.entries(current)) {
      const refVersion = reference[name];
      if (!refVersion) {
        changesDetected.push(`[NEW] ${name}: ${currVersion}`);
      } else if (refVersion !== currVersion) {
        changesDetected.push(`[UPDATED] ${name}: ${refVersion} → ${currVersion}`);
      }
    }

    if (changesDetected.length) {
      await notifyDiscord(changesDetected);
      await saveReferenceVersions(current);
      return res.json({ status: "changes_detected", changes: changesDetected });
    } else {
      return res.json({ status: "no_changes" });
    }
  } catch (err) {
    console.error(err);
    return res.status(500).json({ status: "error", message: err.message });
  }
});

app.get("/", (req, res) => {
  res.json({ message: "Version monitor is running." });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Version monitor listening on port ${PORT}`);
});

File: api/Kdkdjdk.py
import requests
from bs4 import BeautifulSoup
import json
import os
import base64
from datetime import datetime
from flask import Flask, jsonify

app = Flask(__name__)

# Configs — replace with your repo and set GITHUB_TOKEN env var
GITHUB_REPO = "your-username/your-repo"
GITHUB_FILE_PATH = "reference_versions.json"
GITHUB_TOKEN = os.getenv("GITHUB_TOKEN")
DISCORD_WEBHOOK_URL = os.getenv("DISCORD_WEBHOOK_URL")
GITHUB_API_URL = f"https://api.github.com/repos/{GITHUB_REPO}/contents/{GITHUB_FILE_PATH}"

headers = {
    "Authorization": f"token {GITHUB_TOKEN}",
    "Accept": "application/vnd.github+json"
}

TARGET_URL = "https://executors.samrat.lol/"

def fetch_current_versions():
    resp = requests.get(TARGET_URL)
    resp.raise_for_status()
    soup = BeautifulSoup(resp.text, "html.parser")

    executors = {}
    for card in soup.find_all("div", class_="executor-card"):
        name_tag = card.find("h3")
        version_tag = card.find("span", class_="version")
        if name_tag and version_tag:
            name = name_tag.text.strip()
            version = version_tag.text.replace("Version: ", "").strip()
            executors[name] = version
    return executors

def get_file_sha():
    response = requests.get(GITHUB_API_URL, headers=headers)
    if response.status_code == 200:
        return response.json()['sha']
    elif response.status_code == 404:
        return None
    else:
        response.raise_for_status()

def load_reference_versions():
    response = requests.get(GITHUB_API_URL, headers=headers)
    if response.status_code == 200:
        file_content = response.json()['content']
        decoded = base64.b64decode(file_content).decode()
        return json.loads(decoded)
    elif response.status_code == 404:
        return {}
    else:
        response.raise_for_status()

def save_reference_versions(data):
    sha = get_file_sha()
    content = base64.b64encode(json.dumps(data, indent=4).encode()).decode()

    payload = {
        "message": f"Update reference_versions.json at {datetime.utcnow().isoformat()}",
        "content": content,
        "branch": "main",
    }
    if sha:
        payload["sha"] = sha

    response = requests.put(GITHUB_API_URL, headers=headers, json=payload)
    if response.status_code not in [200, 201]:
        raise Exception(f"GitHub update failed: {response.status_code} {response.text}")

def notify_discord(changes):
    if not DISCORD_WEBHOOK_URL:
        print("No Discord webhook URL set. Skipping notification.")
        return
    content = "\n".join(changes)
    payload = {
        "username": "Version Monitor",
        "content": f"**Executor Version Changes Detected:**\n{content}"
    }
    response = requests.post(DISCORD_WEBHOOK_URL, json=payload)
    if not response.ok:
        print(f"Failed to send Discord notification: {response.status_code} {response.text}")

@app.route("/check", methods=["GET"])
def check_versions():
    try:
        reference = load_reference_versions()
        current = fetch_current_versions()

        changes_detected = []

        for name, curr_version in current.items():
            ref_version = reference.get(name)
            if ref_version is None:
                changes_detected.append(f"[NEW] {name}: {curr_version}")
            elif ref_version != curr_version:
                changes_detected.append(f"[UPDATED] {name}: {ref_version} → {curr_version}")

        if changes_detected:
            notify_discord(changes_detected)
            save_reference_versions(current)
            return jsonify({"status": "changes_detected", "changes": changes_detected})
        else:
            return jsonify({"status": "no_changes"})

    except Exception as e:
        return jsonify({"status": "error", "message": str(e)}), 500

@app.route("/", methods=["GET"])
def health():
    return jsonify({"message": "Version monitor is running."})

if __name__ == "__main__":
    app.run(debug=True)
